src/librustc_data_structures/indexed_vec.rs

```diff
@@ -75,23 +92,69 @@ macro_rules! newtype_index {
 
     // Base case, user-defined constants (if any) have already been defined
     (@derives      [$($derives:ident,)*]
-     @pub          [$($pub:tt)*]
      @type         [$type:ident]
      @max          [$max:expr]
+     @vis          [$v:vis]
      @debug_format [$debug_format:tt]) => (
         #[derive(Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, $($derives),*)]
-        pub struct $type($($pub)* u32);
+        $v struct $type {
+            private: ::std::num::NonZeroU32
+        }
+
+        impl $type {
+            $v const MAX_AS_U32: u32 = $max;
+
+            $v const MAX: $type = unsafe { $type::from_u32_unchecked($max) };
+
+            #[inline]
+            $v fn from_usize(value: usize) -> Self {
+                assert!(value <= ($max as usize));
+                unsafe {
+                    $type::from_u32_unchecked(value as u32)
+                }
+            }
+
+            #[inline]
+            $v fn from_u32(value: u32) -> Self {
+                assert!(value <= $max);
```

Comment:
You can make this a `const fn` with some effort, I wish we could enforce the overflow checks on a block of code better. If you really want to make a checked `const fn` variant of this, `assert!(condition)` can become `[()][!condition as usize]`.

Reply:
Ah, cool, I will do that! I was unhappy with the "unchecked" part of this.

Reply:
That said, we did did overflows when I used `u32::MAX` incorrectly...